src/coreclr/inc/corinfo.h

```diff
@@ -2376,6 +2376,15 @@ class ICorStaticInfo
             void **ppIndirection
             ) = 0;
 
+    virtual size_t getIsClassInitedFieldAddress(
```

Comment:
The inlined static cctor trigger is relevant for more than just regular static fields. It is also relevant for thread static fields and static constructor triggers in method prologs. Does the optimization kick for these other cases today?

This JIT/EE method should be about returning information about the inline static constructor trigger. If we need a method that returns a static field address, it should be a different method.

Reply:
So we want to have something like if I understand correctly:
```c
size_t getIsClassInitedFieldAddress(CORINFO_CLASS_HANDLE cls, int32_t* offset);
size_t getStaticBaseAddress(CORINFO_CLASS_HANDLE cls, bool isGC);
```
?

Reply:
I would change `Field` to `Flag` (this is not a managed field) and put the offset last:
```
bool getIsClassInitedFlagAddress(CORINFO_CLASS_HANDLE cls, CORINFO_CONST_LOOKUP* addr, int32_t* offset);
bool getStaticBaseAddress(CORINFO_CLASS_HANDLE cls, bool isGC, CORINFO_CONST_LOOKUP* addr);
```
Looks good otherwise.

> ? (or offset is also not worth

I think offset is good to have.

Are we able to CSE address relocs on Arm64? I am thinking whether it is possible to fetch both the flag and the field value at different offsets from the same base address reloc to reduce the code size on Arm64.

Reply:
Added two new APIs.

> Are we able to CSE address relocs on Arm64? 

Tried to save them to a local, seems that it also improves codegen size on x64 as well.

```csharp
static int s_Int = int.Parse("42");

[MethodImpl(MethodImplOptions.NoInlining)]
int GetInt() => s_Int;
```
NAOT, win-x64:
```asm
; Method UnrollSequenceEqualTests:GetInt():int:this
G_M56961_IG01:              ;; offset=0000H
       56                   push     rsi
       4883EC20             sub      rsp, 32
						;; size=5 bbWeight=1 PerfScore 1.25

G_M56961_IG02:              ;; offset=0005H
       488D3500000000       lea      rsi, [(reloc 0x40000000004216e0)]
       837EF801             cmp      dword ptr [rsi-08H], 1
       7508                 jne      SHORT G_M56961_IG05
						;; size=13 bbWeight=1 PerfScore 4.50

G_M56961_IG03:              ;; offset=0012H
       8B06                 mov      eax, dword ptr [rsi]
						;; size=2 bbWeight=1 PerfScore 2.00

G_M56961_IG04:              ;; offset=0014H
       4883C420             add      rsp, 32
       5E                   pop      rsi
       C3                   ret      
						;; size=6 bbWeight=1 PerfScore 1.75

G_M56961_IG05:              ;; offset=001AH
       E800000000           call     CORINFO_HELP_READYTORUN_NONGCSTATIC_BASE
       EBF1                 jmp      SHORT G_M56961_IG03
						;; size=7 bbWeight=0 PerfScore 0.00
; Total bytes of code: 33
```
NAOT, win-arm64:
```asm
; Method UnrollSequenceEqualTests:GetInt():int:this
G_M56961_IG01:              ;; offset=0000H
        A9BE7BFD          stp     fp, lr, [sp, #-0x20]!
        F9000FF3          str     x19, [sp, #0x18]
        910003FD          mov     fp, sp
						;; size=12 bbWeight=1 PerfScore 2.50

G_M56961_IG02:              ;; offset=000CH
        90000013          adrp    x19, [HIGH RELOC #0x40000000004216e8]
        91000273          add     x19, x19, [LOW RELOC #0x40000000004216e8]
        B85F8260          ldr     w0, [x19, #-0x08]
        7100041F          cmp     w0, #1
        540000A1          bne     G_M56961_IG05
						;; size=20 bbWeight=1 PerfScore 5.50

G_M56961_IG03:              ;; offset=0020H
        B9400260          ldr     w0, [x19]
						;; size=4 bbWeight=1 PerfScore 3.00

G_M56961_IG04:              ;; offset=0024H
        F9400FF3          ldr     x19, [sp, #0x18]
        A8C27BFD          ldp     fp, lr, [sp], #0x20
        D65F03C0          ret     lr
						;; size=12 bbWeight=1 PerfScore 4.00

G_M56961_IG05:              ;; offset=0030H
        94000000          bl      CORINFO_HELP_READYTORUN_NONGCSTATIC_BASE
        17FFFFFB          b       G_M56961_IG03
						;; size=8 bbWeight=0 PerfScore 0.00
; Total bytes of code: 56
```